### ISSUE TITLE
On rust-analyzer nightly e6e72809e, typing colons in format strings panics

### ISSUE THREAD

**Issue by nyanpasu64** [rust-analyzer/rust-analyzer#11303]
**rust-analyzer version**: rust-analyzer version: e6e72809e 2022-01-17 nightly

**rustc version**: rustc 1.57.0 (f1edd0429 2021-11-29)

When I type `format!("");` and add a colon resulting in `format!(":");`, rust-analyzer panics. This also occurs when typing `eprintln!("all remaining input devices:")`. rust-analyzer does not stop working though. I could not reproduce on release 0.2.898.

```
Panic context:
> 
version: e6e72809e 2022-01-17 nightly
request: textDocument/completion CompletionParams {
    text_document_position: TextDocumentPositionParams {
        text_document: TextDocumentIdentifier {
            uri: Url {
                scheme: "file",
                cannot_be_a_base: false,
                username: "",
                password: None,
                host: None,
                port: None,
                path: "/home/nyanpasu64/code/cpal/examples/feedback.rs",
                query: None,
                fragment: None,
            },
        },
        position: Position {
            line: 106,
            character: 43,
        },
    },
    work_done_progress_params: WorkDoneProgressParams {
        work_done_token: None,
    },
    partial_result_params: PartialResultParams {
        partial_result_token: None,
    },
    context: Some(
        CompletionContext {
            trigger_kind: TriggerCharacter,
            trigger_character: Some(
                ":",
            ),
        },
    ),
}

thread '<unnamed>' panicked at 'byte index 3310 is out of bounds of `"all remaining input devices:"`', crates/ide_completion/src/completions/format_string.rs:19:19
stack backtrace:
   0: rust_begin_unwind
             at /rustc/02072b482a8b5357f7fb5e5637444ae30e423c40/library/std/src/panicking.rs:498:5
   1: core::panicking::panic_fmt
             at /rustc/02072b482a8b5357f7fb5e5637444ae30e423c40/library/core/src/panicking.rs:107:14
   2: core::str::slice_error_fail
   3: ide_completion::completions::format_string::format_string
   4: ide_completion::completions
   5: std::panicking::try
   6: rust_analyzer::handlers::handle_completion
   7: std::panicking::try
   8: <F as threadpool::FnBox>::call_box
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
[Error - 9:13:41 PM] Request textDocument/completion failed.
  Message: server panicked: byte index 3310 is out of bounds of `"all remaining input devices:"`
  Code: -32603 
```

**Comment by lnicola** [comment]
Another one, crashing at index 16:

```rust
fn main() {
    let world = "world";
    println!("Hello, {w}")
}
```

3310 is suspiciously large, maybe it's using the wrong offset. I tried to add stuff to the end of the string and it stops crashing but still doesn't offer a completion.

**Comment by nyanpasu64** [comment]
```rust
fn main() {
    ;;;;;;;;;;;;;
    ;;;;;;;;;
    format!(":")}
```

Looks like the wrong index and an integer overflow: `thread '<unnamed>' panicked at 'byte index 4294967295 is out of bounds of `":"`', crates/ide_completion/src/completions/format_string.rs:19:19`

If I add a few more semicolons, it stops panicking, and eventually starts panicking again at low indexes. If I subtract semicolons, the number goes down.

**Comment by Veykril** [comment]
Ye the problem is probably this subtraction here, https://github.com/rust-analyzer/rust-analyzer/blob/e6e72809e3b55da3a57af95e6445a12729331ad6/crates/ide_completion/src/completions/format_string.rs#L15-L17
In tests this doesn't happen since I mocked the `format_args` macro, but for `println` and the like the string token gets lowered and such has different ranges. The subtraction should happen on the original token while the checking for the `format_args` should still be done on the macro downmapped token.

**Comment by lnicola** [comment]
@Veykril do you want to cancel #11304?

**Comment by Veykril** [comment]
#11304 reverted the completions for format strings, so now we need to reenable this and fix the offset problem

**Comment by lnicola** [comment]
Fixed in #11395.
